### DISCUSSION TITLE
Bad file descriptor error with uv_tty_init on Windows

### DISCUSSION THREAD

**Discussion by Aidan63** [libuv/libuv#3824] [Q&A] [answered]
Hello,

I've been running into an issue where `uv_tty_init` returns the bad file descriptor error on Windows. For some context this program is a debugger which is launched by vscode and it only occurs when vscode attempts to launch my debugger, if I manually launch it then opening the tty succeeds.
My initial thoughts was that vscode was spawning the process without a console so that was causing the error, however this doesn't appear to be true. Looking at the win.c file it ultimately calls `_get_osfhandle` to get the handle, calling this myself I can get handles for stdin, out, and err and none of them are invalid.

```c++
    auto hStdin  = reinterpret_cast<HANDLE>(_get_osfhandle(0)) == INVALID_HANDLE_VALUE ? std::wstring(L"invalid") : std::wstring(L"valid");
    auto hStdout = reinterpret_cast<HANDLE>(_get_osfhandle(1)) == INVALID_HANDLE_VALUE ? std::wstring(L"invalid") : std::wstring(L"valid");
    auto hStderr = reinterpret_cast<HANDLE>(_get_osfhandle(2)) == INVALID_HANDLE_VALUE ? std::wstring(L"invalid") : std::wstring(L"valid");

    MessageBoxW(
        nullptr,
        (hStdin + std::wstring(L" : ") + hStdout + std::wstring(L" : ") + hStdout).c_str(),
        std::wstring(L"test").c_str(),
        MB_OK)
```
(this only occuring when launching from vscode makes debugging a pain, hence the MessageBox call...)

I also tried `AllocConsole` and this fails with access is denied meaning a console is already attached. I've also confirmed that the handle initially returned within uv_tty_init isn't invalid, as the `INVALID_HANDLE_VALUE` check doesn't return the UV_EBADF there, so it must come from somewhere else. Anyone have any thoughts? This only occuring when the program is launched from vscode makes debugging a slog, so I was hoping someone might have some pointers.

When I compile the program on linux all the tty stuff works fine when vscode launches it.

Thanks.

**Comment by bnoordhuis** [comment, marked as answer]
> I've been running into an issue where uv_tty_init returns the bad file descriptor error on Windows.

You mean `uv_tty_init()` returns `UV_EBADF`? Just spitballing but is it possible the `DuplicateHandle()` call fails somehow?

https://github.com/libuv/libuv/blob/64669fdd8de8fce478e6ce1f7c954905f682c75b/src/win/tty.c#L198-L214

edit: you can probably track down the failing system call with printf-style debugging but that means building node from source. The relevant file is `deps/uv/src/win/tty.c`.

**Reply by Aidan63** [reply]
Thanks, yes, `uv_tty_init` returns `UV_EBADF`. I did some more debugging this morning and managed to get visual studio to attach to the process vscode launches so no more MessageBox debugging, and I've managed to figure it out.

`DuplicateHandle` wasn't failing but the following `GetNumberOfConsoleInputEvents` returns not readable and that block is entered. `GetConsoleScreenBufferInfo` was the function which was failing. Unfortunately my attempts at bodging to code to force the readable path also didn't work as the handles did not have the appropriate permissions for reading and writing.

In googling around I saw some references to named pipes for vscode's debugger communication and decided to try opening and reading from stdin with a uv_pipe and it worked! Why tty worked fine on linux is something I still don't know (and need to make sure this pipe stuff works there as well).
